python/python/tests/test_pubsub.py

```diff
@@ -0,0 +1,1800 @@
+
+        # assert there are no messages to read
+        with pytest.raises(asyncio.TimeoutError):
+            await asyncio.wait_for(listening_client.get_pubsub_message(), timeout=3)
+
+        assert listening_client.try_get_pubsub_message() is None
+
+        if cluster_mode:
+            # Since all tests run on the same cluster, when closing the client, garbage collector can be called after another test will start running
+            # In cluster mode, we check how many subscriptions received the message
+            # So to avoid flakiness, we make sure to unsubscribe from the channels
+            await listening_client.custom_command(
+                ["UNSUBSCRIBE", *list(channels_and_messages.keys())]
+            )
+
+    @pytest.mark.parametrize("cluster_mode", [True])
+    @pytest.mark.parametrize(
+        "method", [MethodTesting.Async, MethodTesting.Sync, MethodTesting.Callback]
+    )
+    async def test_sharded_pubsub(
+        self, request, cluster_mode: bool, method: MethodTesting
+    ):
+        """
+        Test sharded Pub/Sub functionality with different message retrieval methods.
+
+        This test covers the sharded pub/sub flow using three different methods:
+        Async, Sync, and Callback. It verifies that a message published to a
+        specific sharded channel is correctly received by a subscriber.
+        """
+        channel = get_random_string(10)
+        message = get_random_string(5)
+        publish_response = 1
+
+        callback, context = None, None
+        callback_messages: List[
+            Union[RedisClient.PubSubMsg, RedisClusterClient.PubSubMsg]
+        ] = []
+        if method == MethodTesting.Callback:
+            callback = new_message
+            context = callback_messages
+
+        pub_sub = create_pubsub_subscription(
+            cluster_mode,
+            {PubSubChannelModes.Sharded: {channel}},
+            callback=callback,
+            context=context,
+        )
+
+        publishing_client, listening_client = await setup_clients(
+            request, cluster_mode, pub_sub
+        )
+        min_version = "7.0.0"
+        if await check_if_server_version_lt(publishing_client, min_version):
+            pytest.skip(reason=f"Redis version required >= {min_version}")
+
+        assert type(publishing_client) == RedisClusterClient
+        assert (
+            await publishing_client.publish(message, channel, sharded=True)
+            == publish_response
+        )
+        # allow the message to propagate
+        await asyncio.sleep(1)
+
+        pubsub_msg = await get_message_by_method(
+            method, listening_client, callback_messages, 0
+        )
+        assert pubsub_msg.message == message
+        assert pubsub_msg.channel == channel
+        assert pubsub_msg.pattern is None
+
+        # assert there are no messages to read
+        await check_no_messages_left(method, listening_client)
+        if cluster_mode:
```

Comment:
This is always true in this test.
Meanwhile you can move this common piece of code to an auxiliary function.
I like that you described uneven situations.